Implement GPU skinning

setzer22 opened the pull request:
<!-- Thank you for making a pull request! Below are the recommended steps to validate/complete your PR -->

## Checklist

- CI Checked:
  - [x] `cargo fmt` has been ran
  - [x] `cargo clippy` reports no issues
  - [x] `cargo test` succeeds
  - [x] `cargo rend3-doc` has no warnings
  - [x] `cargo deny check` issues have been fixed or added to deny.toml
- Manually Checked:
  - [x] relevant examples/test cases run
  - [x] changes added to changelog
    - [x] Add credit to yourself for each change: `Added new functionality @githubname`.

## Related Issues
Closes #27

## Description
This brings the remaining pieces to make the skinning system work. Previous PRs: #326, #325

This PR adds the following features:
- Adds pre-skinning (data upload) and skinning (compute pass) nodes to the default rendergraph.
- Fixes `copy_buffer_to_buffer` not working with same src/dst buffers by using a compute shader instead (see `buffer_copier.rs`).
- Adds a new `skinning` example.
- Adds autodetection of skinned meshes in rend3-gltf. A skeleton is now created for those meshes instead.

setzer22 committed:
Add code for pre-skinning. Rename to joint_matrices.

setzer22 committed:
Add pre skinning to rend3 routine

setzer22 committed:
Load skins in rend3_gltf

setzer22 committed:
Autodetect and spawn skinned meshes in rend3_gltf

setzer22 committed:
Introduce buffer copier to workaround copy_buffer_to_buffer issue

setzer22 committed:
Add new skinning example

setzer22 committed:
Add GPU skinner and compute pass boilerplate

setzer22 committed:
Add GPU skinning shader type declarations.

Also:
- Make sure the nodes in the graph run.
- Fix minor layout / BGL issues.

setzer22 committed:
Add utility method to debug compute shader

setzer22 committed:
Implement skinning shader

setzer22 committed:
Merge latest trunk, resolve conflicts

setzer22 committed:
Resolve clippy issues, run cargo fmt

setzer22 committed:
Fix missing link in doc

setzer22 committed:
Improve the skinning example. Add 'Armature' type on rend3-gltf.

setzer22 committed:
Replace copy of inv bind matrices vector with a skin index

setzer22 committed:
Add missing screenshot

setzer22 committed:
Do not run skinning when there are no skeletons

Prevents binding an empty buffer, which causes a panic.

setzer22 commented on examples/skinning/src/lib.rs:
```diff
@@ -0,0 +1,170 @@
+    directional_light_handle: Option<rend3::types::DirectionalLightHandle>,
+}
+
+/// Locates an object in the node hierarchy that corresponds to an animated mesh
+/// and returns its list of skeletons. Note that a gltf object may contain
+/// multiple primitives, and there will be one skeleton per primitive.
+pub fn find_armature<'a>(
+    nodes: impl Iterator<Item = &'a rend3_gltf::Labeled<rend3_gltf::Node>>,
+) -> Option<rend3_gltf::Armature> {
+    for node in nodes {
+        if let Some(ref obj) = node.inner.object {
+            if let Some(ref armature) = obj.inner.armature {
+                return Some(armature.clone());
+            }
+        }
+        if let Some(skels) = find_armature(node.inner.children.iter()) {
+            return Some(skels);
+        }
+    }
+    None
+}
```
I think there is potential to add more functions like this to rend3-gltf, but I'll leave that for a later PR.

setzer22 commented on rend3-gltf/src/lib.rs:
```diff
@@ -65,13 +65,29 @@ pub struct Mesh {
     pub primitives: Vec<MeshPrimitive>,
 }
 
+/// A set of [`SkeletonHandle`]s, one per mesh in the wrapping object, plus the
+/// index of the skin data in the `skins` array of [`LoadedGltfScene`].
+///
+/// All the skeletons are guaranteed to have the same number of joints and
+/// structure, but deform different meshes of the object.
+#[derive(Debug, Clone)]
+pub struct Armature {
+    /// The list of skeletons that are deforming this node's mesh primitives.
+    pub skeletons: Vec<SkeletonHandle>,
+    /// Index to the skin that contains the inverse bind matrices for the
+    /// skeletons in this armature.
+    pub skin_index: usize,
+}
+
```
I chose `Armature` for lack of a better name. But essentially this is a struct that groups all of the `Skeleton`s in a single logical thing when there's a gltf node that has multiple skinned primitives.

setzer22 commented on rend3-gltf/src/lib.rs:
```diff
@@ -102,6 +118,10 @@ pub struct Texture {
     pub format: types::TextureFormat,
 }
 
+#[derive(Debug)]
+pub struct Skin {
+    pub inverse_bind_matrices: Vec<Mat4>,
+}
```
The skin also has joint information we probably want to expose later. For now I still haven't figured out how to interpret that data, so we can decide how to best expose it in another PR

setzer22 commented on rend3-gltf/src/lib.rs:
```diff
@@ -312,35 +337,60 @@ pub fn add_mesh_by_index<E: std::error::Error + 'static>(
+        None
+    };
+
+    for prim in mesh_handle.inner.primitives.iter() {
```
The code here became a bit too involved for a nice iterator chain, so I refactored it into a good ol' imperative loop that pushes to vectors instead.

cwfitzgerald replied:
nit: 
```suggestion
    for prim in &mesh_handle.inner.primitives {
```

setzer22 commented on rend3-routine/src/skinning/mod.rs:
```diff
@@ -0,0 +1,272 @@
+            push_constant_ranges: &[],
+        });
+
+        let module = device.create_shader_module(&wgpu::ShaderModuleDescriptor {
+            label: Some("Gpu skinning compute shader"),
+            source: wgpu::ShaderSource::Wgsl(include_str!("../../shaders/src/skinning.wgsl").into()),
+        });
```
I'm not sure how the shader loading mechanism in rend3-routine works, and to be honest I'm also not very familiar with the specific flavor of GLSL that Vulkan/Wgpu use. So I went for a wgsl shader instead.

cwfitzgerald replied:
This kind of steady transition is what's going to work well to transition, so I'm glad you're starting it!

setzer22 commented on rend3-routine/shaders/src/skinning.wgsl:
```diff
@@ -0,0 +1,110 @@
+        return;
+    }
+
+    let pos = to_v(positions.data[input.mesh_range.start + idx]);
+    let normal = to_v(normals.data[input.mesh_range.start + idx]);
+    let tangent = to_v(tangents.data[input.mesh_range.start + idx]);
+    let joint_ws = joint_weights.data[input.mesh_range.start + idx];
+    let joint_is = joint_indices.data[input.mesh_range.start + idx];
+
+    let joint_i0 = joint_is.indices_0_1 & 0x0000ffffu;
+    let joint_i1 = (joint_is.indices_0_1 & 0xffff0000u) >> 16u;
+    let joint_i2 = joint_is.indices_2_3 & 0x0000ffffu;
+    let joint_i3 = (joint_is.indices_2_3 & 0xffff0000u) >> 16u;
+    
+    // Compute the skinned position and write to output region of buffer
+    let pos4 = vec4<f32>(pos, 1.0);
+    var pos_acc = vec3<f32>(0.0, 0.0, 0.0);
+    pos_acc = pos_acc + ((pos4 * get_joint_matrix(joint_i0)) * joint_ws.w0).xyz;
+    pos_acc = pos_acc + ((pos4 * get_joint_matrix(joint_i1)) * joint_ws.w1).xyz;
+    pos_acc = pos_acc + ((pos4 * get_joint_matrix(joint_i2)) * joint_ws.w2).xyz;
+    pos_acc = pos_acc + ((pos4 * get_joint_matrix(joint_i3)) * joint_ws.w3).xyz;
+    positions.data[input.skeleton_range.start + idx] = from_v(pos_acc);
+
+    // Compute the skinned normal and write to output region of buffer
+    let norm4 = vec4<f32>(normal, 0.0);
+    var norm_acc = vec3<f32>(0.0, 0.0, 0.0);
+    norm_acc = norm_acc + ((norm4 * get_joint_matrix(joint_i0)) * joint_ws.w0).xyz;
+    norm_acc = norm_acc + ((norm4 * get_joint_matrix(joint_i1)) * joint_ws.w1).xyz;
+    norm_acc = norm_acc + ((norm4 * get_joint_matrix(joint_i2)) * joint_ws.w2).xyz;
+    norm_acc = norm_acc + ((norm4 * get_joint_matrix(joint_i3)) * joint_ws.w3).xyz;
+    normals.data[input.skeleton_range.start + idx] = from_v(norm_acc);
+
+    // Compute the skinned tangent and write to output region of buffer
+    let tang4 = vec4<f32>(tangent, 0.0);
+    var tang_acc = vec3<f32>(0.0, 0.0, 0.0);
+    tang_acc = tang_acc + ((tang4 * get_joint_matrix(joint_i0)) * joint_ws.w0).xyz;
+    tang_acc = tang_acc + ((tang4 * get_joint_matrix(joint_i1)) * joint_ws.w1).xyz;
+    tang_acc = tang_acc + ((tang4 * get_joint_matrix(joint_i2)) * joint_ws.w2).xyz;
+    tang_acc = tang_acc + ((tang4 * get_joint_matrix(joint_i3)) * joint_ws.w3).xyz;
+    tangents.data[input.skeleton_range.start + idx] = from_v(tang_acc);
```
Would love to use a loop for this, but it's probably more trouble than it's worth. But if you know of a more elegant way of doing this please let me know! :smile:

cwfitzgerald replied:
If you make JointWeightVec a array<T, 4> you should be able to iterate over it.

setzer22 replied:
~~Doesn't look like I can :(~~  *EDIT:* Fixed!

```
Shader validation error: 
   ┌─ Gpu skinning compute shader:96:26
   │
96 │         let joint_index = joint_indices[i];
   │                          ^^^^^^^^^^^^^^^^^ naga::Expression [103]


    Entry point main at Compute is invalid
    Expression [103] is invalid
    The expression [72] may only be indexed by a constant
```
I asked on the wgpu chat, but I'll keep it like this for now

setzer22 commented on rend3/src/managers/skeleton.rs:
```diff
@@ -48,34 +69,40 @@ impl SkeletonManager {
+        let mesh_range = mesh_manager.internal_data(skeleton.mesh.get_raw()).vertex_range.clone();
+        let skeleton_range = mesh_manager.allocate_skeleton_mesh(device, queue, encoder, &skeleton.mesh);
+
+        let input = GpuVertexRanges {
+            skeleton_range: [skeleton_range.start as u32, skeleton_range.end as u32],
+            mesh_range: [mesh_range.start as u32, mesh_range.end as u32],
+        };
```
I followed a similar strategy to culling, where this is cached on the `InternalSkeleton` to avoid having to fetch the mesh and its range in a hot loop (during pre-skinning).

cwfitzgerald replied:
This is the right way to go about it, but we also need to properly invalidate it when vertex buffers reallocate. #332 is just this but with objects and indices.

cwfitzgerald commented on examples/skinning/Cargo.toml:
```diff
@@ -0,0 +1,54 @@
+# windowing
+winit = "0.26"
+# block on async functions
+pollster = "0.2"
```
Nit: we should sort these alphabetically.

cwfitzgerald commented on examples/skinning/src/lib.rs:
```diff
@@ -0,0 +1,170 @@
+    nodes: impl Iterator<Item = &'a rend3_gltf::Labeled<rend3_gltf::Node>>,
+) -> Option<rend3_gltf::Armature> {
+    for node in nodes {
+        if let Some(ref obj) = node.inner.object {
+            if let Some(ref armature) = obj.inner.armature {
+                return Some(armature.clone());
+            }
+        }
```
(not a problem) I really wish rust had a `?.` that carried forward options

setzer22 replied:
Me too! `and_then` and `map` are nice, but often not much better than this

cwfitzgerald commented on rend3-routine/shaders/src/skinning.wgsl:
```diff
@@ -0,0 +1,110 @@
+    let joint_ws = joint_weights.data[input.mesh_range.start + idx];
+    let joint_is = joint_indices.data[input.mesh_range.start + idx];
+
+    let joint_i0 = joint_is.indices_0_1 & 0x0000ffffu;
```
Could you drop a note that this should use `bitfieldExtract` once it actually works on dx12

cwfitzgerald commented on rend3-routine/shaders/src/skinning.wgsl:
```diff
@@ -0,0 +1,110 @@
+    // Compute the skinned position and write to output region of buffer
+    let pos4 = vec4<f32>(pos, 1.0);
+    var pos_acc = vec3<f32>(0.0, 0.0, 0.0);
+    pos_acc = pos_acc + ((pos4 * get_joint_matrix(joint_i0)) * joint_ws.w0).xyz;
+    pos_acc = pos_acc + ((pos4 * get_joint_matrix(joint_i1)) * joint_ws.w1).xyz;
+    pos_acc = pos_acc + ((pos4 * get_joint_matrix(joint_i2)) * joint_ws.w2).xyz;
+    pos_acc = pos_acc + ((pos4 * get_joint_matrix(joint_i3)) * joint_ws.w3).xyz;
```
How does this deal with situations where there is < 4 matrices? We should _probably_ be conditionally skipping those cases considering how matrix multiplication and the memory read will hurt.

cwfitzgerald replied:
The right multiplication of the matrix seems wrong here, I would expect this to be `(mat * pos) * weight`

setzer22 replied:
> The right multiplication of the matrix seems wrong here, I would expect this to be (mat * pos) * weight

You're right, it was. It's fixed now. I also updated the skinning example to make it a bit more interesting while I was debugging this

setzer22 replied:
> How does this deal with situations where there is < 4 matrices?

The new code now only reads the matrix if the weight is greater than `0.0`

cwfitzgerald commented on rend3-routine/shaders/src/skinning.wgsl:
```diff
@@ -0,0 +1,110 @@
+    // Compute the skinned normal and write to output region of buffer
+    let norm4 = vec4<f32>(normal, 0.0);
+    var norm_acc = vec3<f32>(0.0, 0.0, 0.0);
+    norm_acc = norm_acc + ((norm4 * get_joint_matrix(joint_i0)) * joint_ws.w0).xyz;
+    norm_acc = norm_acc + ((norm4 * get_joint_matrix(joint_i1)) * joint_ws.w1).xyz;
+    norm_acc = norm_acc + ((norm4 * get_joint_matrix(joint_i2)) * joint_ws.w2).xyz;
+    norm_acc = norm_acc + ((norm4 * get_joint_matrix(joint_i3)) * joint_ws.w3).xyz;
```
This won't work for non-uniform scales, you need to multiply the input vector by the squared scale of the matrix. See https://github.com/BVE-Reborn/rend3/blob/trunk/rend3-routine/shaders/src/opaque.vert#L55 and https://github.com/BVE-Reborn/rend3/blob/trunk/rend3-routine/src/culling/cpu.rs#L98-L102. i.e. you can get the squared scale with 3 dot products. `vec3(dot(mat[0].xyz, mat[0].xyz), dot(mat[1].xyz, mat[1].xyz), dot(mat[2].xyz, mat[2].xyz))`

setzer22 replied:
> you need to multiply the input vector by the squared scale of the matrix
I looked at the code and I saw it uses `inv_squared_scale`, so I assumed you meant the *inverse* of the squared scale.

I think I got it right, but please double-check! :+1:

cwfitzgerald commented on rend3-routine/src/skinning/mod.rs:
```diff
@@ -0,0 +1,272 @@
+        mapped_at_creation: true,
+    });
+
+    let joint_matrices_length: usize = skeleton_manager
+        .skeletons()
+        .map(|sk| sk.joint_matrices.len())
+        .sum::<usize>()
```
We should statefully keep this number inside the skeleton manager to avoid needing a full pass over all skeletons.

setzer22 replied:
Took care of this. There's now a new `global_joint_count` field inside the `SkeletonManager` that's used instead. I believe I took care of updating that count at all the relevant places:

- In `SkeletonManager::fill`, incrementing by the necessary amount
- During `ready`, in `remove_all_dead`, where the amount is subtracted

cwfitzgerald commented on rend3-routine/src/skinning/mod.rs:
```diff
@@ -0,0 +1,272 @@
+                wgpu::BindingType::Buffer {
+                    ty: wgpu::BufferBindingType::Uniform,
+                    has_dynamic_offset: true,
+                    min_binding_size: None,
```
We should set min binding sizes as this removes draw time work for wgpu

setzer22 replied:
I believe I did that right, using the size of the `GpuSkinningInput` for the uniform buffer, and the size of a single element for the vertex buffers. But please double check just in case

cwfitzgerald commented on rend3-routine/src/skinning/mod.rs:
```diff
@@ -0,0 +1,272 @@
+        skeleton_manager: &SkeletonManager,
+    ) {
+        let vertex_buffers_bg = {
+            let mut bgb = BindGroupBuilder::new();
+            bgb.append_buffer(&mesh_buffers.vertex_position);
+            bgb.append_buffer(&mesh_buffers.vertex_normal);
+            bgb.append_buffer(&mesh_buffers.vertex_tangent);
+            bgb.append_buffer(&mesh_buffers.vertex_joint_index);
+            bgb.append_buffer(&mesh_buffers.vertex_joint_weight);
+            bgb.build(device, Some("GPU skinning vertex buffers"), &self.vertex_buffers_bgl)
```
Super duper nit: these can be chained, so `BindGroupBuilder::new().append_buffer(...).build()` and we don't need the mut or the block here

setzer22 replied:
Much nicer, thanks! :smile:

cwfitzgerald commented on rend3-routine/src/skinning/mod.rs:
```diff
@@ -0,0 +1,272 @@
+        };
+
+        for (i, skel) in skeleton_manager.skeletons().enumerate() {
+            let mut cpass = encoder.begin_compute_pass(&wgpu::ComputePassDescriptor {
+                label: Some(&format_sso!("GPU Skinning {}", i)),
+            });
```
We should lift this cpass outside of the loop -- passes carry some overhead, so we should minimize the count. It also means we can bind the bind groups only once.

cwfitzgerald commented on rend3-routine/src/skinning/mod.rs:
```diff
@@ -0,0 +1,272 @@
+            cpass.set_bind_group(2, &skinning_inputs_bgb, &[offset]);
+
+            let num_verts = (skel.ranges.mesh_range[1] - skel.ranges.mesh_range[0]) as u32;
+            let num_workgroups = (num_verts + (Self::WORKGROUP_SIZE - 1)) / Self::WORKGROUP_SIZE; // Divide rounding up
```
Could we add a function that deals with this right next to `rend3::util::math::round_up_pot` https://github.com/BVE-Reborn/rend3/blob/trunk/rend3/src/util/math.rs#L3-L7

setzer22 replied:
I was just looking for a good place to add this, especially since I had to write this twice on two separate places, and most compute shaders will likely want to do integer divide rounding up :+1:

cwfitzgerald commented on rend3-routine/src/skinning/mod.rs:
```diff
@@ -0,0 +1,272 @@
+    graph: &mut RenderGraph<'node>,
+    gpu_skinner: &'node GpuSkinner,
+    pre_skin_data: DataHandle<PreSkinningBuffers>,
+    skinned_data: DataHandle<()>,
```
We should give this an empty tag type:

```
struct SkinningOutput;
```

which is about equivalent and as we collect more of these pieces of dummy data will prevent putting the wrong data in the wrong hole.

cwfitzgerald commented on rend3/src/managers/skeleton.rs:
```diff
@@ -12,9 +12,30 @@ use wgpu::{CommandEncoder, Device, Queue};
+#[derive(Debug, Copy, Clone)]
+pub struct GpuVertexRanges {
+    /// The range of the vertex buffer that holds the original mesh.
+    pub mesh_range: [u32; 2],
```
Should probably be glam UVec2

cwfitzgerald commented on rend3/src/util/buffer_copier.rs:
```diff
@@ -0,0 +1,142 @@
+            cpass.dispatch(num_workgroups, 1, 1);
+        }
+
+        queue.submit(Some(encoder.finish()));
```
We shouldn't submit here -- we should passthrough this encoder: https://github.com/BVE-Reborn/rend3/blob/trunk/rend3/src/renderer/ready.rs#L18-L20 which will automatically get submitted before everything else.

cwfitzgerald reviewed (changes requested):
This is really good stuff! I'm very glad you're tackling this and so well.

Lots of comments, but nothing fundamentally broken

cwfitzgerald commented on rend3/Cargo.toml:
```diff
@@ -57,3 +57,4 @@ smartstring = "0.2"
 thiserror = "1"
 wgpu = "0.12"
 wgpu-profiler = "0.8.0"
+pollster = "0.2"
```
Seems to be unused, and we can't use it on wasm anyway.

setzer22 committed:
Address some review comments

setzer22 committed:
Improve skinning example. Transform shader into a loop.

setzer22 committed:
Fix normals / tangents computation

cwfitzgerald commented on examples/skinning/src/lib.rs:
```diff
@@ -146,7 +165,7 @@ impl rend3_framework::App for SkinningExample {
                     resolution,
                     SAMPLE_COUNT,
-                    glam::Vec4::ZERO,
+                    glam::Vec4::ONE * 0.15,
```
```suggestion
                    glam::Vec4::splat(0.15),
```

cwfitzgerald commented on rend3-routine/shaders/src/skinning.wgsl:
```diff
@@ -73,38 +81,38 @@ fn main([[builtin(global_invocation_id)]] global_id: vec3<u32>) {
-    tangents.data[input.skeleton_range.start + idx] = from_v(tang_acc);
+    
+    for (var i = 0; i < 4; i = i + 1) {
+        let joint_index = joint_indices[i];
```
Out of an abundance of caution, could we move this inside the if statement.

cwfitzgerald commented:
Apparently I filled out the comments but never hit go yesterday, sorry!

setzer22 commented:
> Apparently I filled out the comments but never hit go yesterday, sorry!

No worries! Will get around to finishing the remaining comments on monday :)

setzer22 committed:
Address more review comments.

setzer22 committed:
Lift cpass outside of loop and add math::round_up_div

setzer22 committed:
Make skinning output use empty tag instead of ()

setzer22 committed:
Use the primary encoder instead of creating a new one.

setzer22 committed:
Represent GpuVertexRanges as UVec2 instead of [u32;2]

setzer22 committed:
Store back-references to skeletons in InternalMesh

setzer22 committed:
Add min_binding_size for buffers in the gpu skinner

setzer22 committed:
Formatting

setzer22 committed:
Fix more clippy lints

setzer22 committed:
Merge remote-tracking branch 'upstream/trunk' into feat/gpu_skinning

And resolve merge conflicts

setzer22 committed:
Implement fix for #332 with the new skeleton data

- The skeleton data is now copied during reallocation
- The ranges stored in GpuVertexRanges inside the skeleton are now
  updated when that happens

setzer22 committed:
Allow clippy lint on mesh manager fill()

cwfitzgerald commented on rend3-routine/src/skinning/mod.rs:
```diff
@@ -46,14 +54,9 @@ fn build_gpu_skinning_input_buffers(device: &Device, skeleton_manager: &Skeleton
     let joint_matrices = device.create_buffer(&BufferDescriptor {
         label: Some("joint matrices"),
-        size: joint_matrices_length as u64,
+        size: (skeleton_manager.global_joint_count() * std::mem::size_of::<[[f32; 4]; 4]>()) as u64,
```
You can just use `Mat4` here.

nit: as a general style thing, I do `mem::size_of` and import `std::mem` to make it a bit easier to type.

cwfitzgerald commented on rend3-routine/src/skinning/mod.rs:
```diff
@@ -117,50 +120,51 @@ impl GpuSkinner {
+        let tan_size = NonZeroU64::new(VERTEX_TANGENT_SIZE as u64);
+        let j_idx_size = NonZeroU64::new(VERTEX_JOINT_INDEX_SIZE as u64);
+        let j_wt_size = NonZeroU64::new(VERTEX_JOINT_WEIGHT_SIZE as u64);
+        let mat_size = NonZeroU64::new(std::mem::size_of::<[[f32; 4]; 4]>() as u64);
```
Mat4 here too

cwfitzgerald commented on rend3/src/managers/skeleton.rs:
```diff
@@ -91,13 +100,21 @@ impl SkeletonManager {
 
     pub fn set_joint_matrices(&mut self, handle: RawSkeletonHandle, joint_matrices: Vec<Mat4>) {
         let skeleton = self.registry.get_mut(handle);
+        assert_eq!(joint_matrices.len(), skeleton.joint_matrices.len());
```
Let's add a error message here of some kind:
```suggestion
        assert_eq!(joint_matrices.len(), skeleton.joint_matrices.len(), "Joint matrix count cannot be changed");
```

Not sure that's the best message, but you get the picture.

cwfitzgerald commented on rend3/src/util/math.rs:
```diff
@@ -5,3 +5,8 @@ pub fn round_up_pot<T: PrimInt>(src: T, factor: T) -> T {
     (src + minus1) & !minus1
 }
+
+/// Performs integer division betwee a and b rounding up, instead of down
```
```suggestion
/// Performs integer division between a and b rounding up, instead of down
```
😄

cwfitzgerald commented on rend3-routine/src/skinning/mod.rs:
```diff
@@ -197,50 +201,49 @@ impl GpuSkinner {
     }
 }
 
+pub struct SkinningOutput;
```
Docs

cwfitzgerald commented on rend3-routine/src/skinning/mod.rs:
```diff
@@ -197,50 +201,49 @@ impl GpuSkinner {
 
+pub struct SkinningOutput;
+
 pub fn add_skinning_to_graph<'node>(
```
Docs

cwfitzgerald commented on rend3-routine/src/skinning/mod.rs:
```diff
@@ -117,50 +120,51 @@ impl GpuSkinner {
                     ty: wgpu::BufferBindingType::Uniform,
                     has_dynamic_offset: true,
-                    min_binding_size: None,
+                    min_binding_size: NonZeroU64::new(std::mem::size_of::<GpuSkinningInput>() as u64),
```
```suggestion
                    min_binding_size: NonZeroU64::new(mem::size_of::<GpuSkinningInput>() as u64),
```

cwfitzgerald commented on examples/skinning/README.md:
```diff
@@ -0,0 +1,5 @@
+
+This example showcases how to import a mesh with an armature and positions its bones.
+
+![](screenshot.png)
```
Don't forget to add this to the examples readme

cwfitzgerald commented on rend3-gltf/src/lib.rs:
```diff
@@ -312,35 +337,60 @@ pub fn add_mesh_by_index<E: std::error::Error + 'static>(
-        })
-        .collect();
+
+    let mut primitives: Vec<ObjectHandle> = vec![];
+    let mut skeletons: Vec<SkeletonHandle> = vec![];
```
```suggestion
    let mut primitives = Vec:;new();
    let mut skeletons = Vec::new();
```

I feel like the type annotations shouldn't be needed, but if they are, ignore this suggestion.

cwfitzgerald reviewed (changes requested):
We're getting pretty nitpicky now, so hopefully this is the last round :)

Nothing ever calls `SkeletonManager::ready`, which should be called by the ready function of the renderer.

Super excited for this to land!

setzer22 committed:
betwee -> between

Co-authored-by: Connor Fitzgerald <connorwadefitzgerald@gmail.com>

setzer22 committed:
Another round of review comments

setzer22 committed:
Ensure vertex joint indices are in bounds.

setzer22 committed:
Cargo fmt && cargo clippy

cwfitzgerald commented on rend3/src/managers/mesh.rs:
```diff
@@ -153,17 +178,14 @@ impl MeshManager {
-                objects: Vec::new(),
-            };
-
+            let mesh = InternalMesh::new_empty();
```
Good call

setzer22 committed:
Call SkeletonManager::ready()

Also fix a bug where the owned region of the vertex buffer was
not freed when a skeleton was removed.

cwfitzgerald reviewed (approved):
Absolutely amazing work!

setzer22 commented:
Thanks! :smile: :tada:

scoopr commented on rend3/src/managers/skeleton.rs:
```diff
@@ -46,41 +72,96 @@ impl SkeletonManager {
+            "Call to set_joint_matrices with an incorrect number of bones. \
+            Skeleton has {} bones, input vector has {}.",
+            joint_matrices.len(),
+            skeleton.joint_matrices.len(),
```
I think the formatting arguments are wrong way around
